### PR TITLE
Add Pin::static_ref, static_mut.

### PR DESCRIPTION
**Pull request by m-ou-se**
This adds `Pin::static_ref` and `Pin::static_mut`, which convert a static reference to a pinned static reference.

Static references are effectively already pinned, as what they refer to has to live forever and can never be moved.

---

Context: I want to update the `sys` and `sys_common` mutexes/rwlocks/condvars to use `Pin<&self>` in their functions, instead of only warning in the unsafety comments that they may not be moved. That should make them a little bit less dangerous to use. Putting such an object in a `static` (e.g. through `sys_common::StaticMutex`) fulfills the requirements about never moving it, but right now there's no safe way to get a `Pin<&T>` to a `static`. This solves that.

### CODE REVIEW

**Comment by withoutboats** [comment]
Separate from my comments on #77801, I think this method is somewhat interesting. I can think of a use case of using it to pin a [`PinCell`](https://docs.rs/pin-cell/0.1.1/pin_cell/struct.PinCell.html) for example.

I think the mutable equivalent is even more interesting, though, because other than PinCell there is not much use for `Pin<&T>`. So I would add these unstably as a pair of methods: `static_ref` and `static_mut`:

```rust
impl<T: ?Sized> Pin<&'static T> {
    pub fn static_ref(r: &'static T) -> Self { ... }
}

impl<T: ?Sized> Pin<&'static mut T> {
    pub fn static_mut(r: &'static mut T) -> Self { ... }
}
```

File: library/core/src/pin.rs [hunk 1]
```diff
@@ -781,6 +781,34 @@ impl<'a, T: ?Sized> Pin<&'a mut T> {
...
+    #[unstable(feature = "pin_static_ref", issue = "none")]
+    #[rustc_const_unstable(feature = "const_pin", issue = "76654")]
+    pub const fn static_ref(r: &'static T) -> Pin<&'static T> {
+        // SAFETY: The 'static borrow guarantees the data will not be
+        // moved/invalidated until it gets dropped (which is never).
+        unsafe { Pin::new_unchecked(r) }
+    }
+}
+
+impl<T: ?Sized> Pin<&'static T> {
```
Comment (dtolnay) [inline library/core/src/pin.rs:798]:
```suggestion
impl<T: ?Sized> Pin<&'static mut T> {
```

**Comment by withoutboats** [comment]
@rfcbot fcp merge

This PR adds two constructors for pinned references: from `&'static T` and `&'static mut T`; these constructors are sound because you can never move or invalidate something to which you have a `'static` reference.

**Comment by withoutboats** [comment]
Another API we could consider in the future (but I'm not sure its useful enough) is adding `leak` to `Pin<Box<T>>` etc:

```rust
impl<T: ?Sized> Pin<Box<T>> {
    fn leak(this: Pin<Box<T>>) -> Pin<&'static mut T> { /* .. */ }
}

// equivalent expressions:
let x = Pin::leak(Box::pin(expr));
let x = Pin::static_mut(Box::leak(Box::new(expr)));
```

**Review by dtolnay — APPROVED** [review]
I see 3/6 checkboxes and this PR is just adding unstable methods so I think we can land without waiting for an FCP.

**Comment by Mark-Simulacrum** [comment]
This was backported into 1.48 (approved by me) in #79075 as a dependency of #77801.

**Pull request merged**